### ISSUE TITLE
holland 1.1.5 won't connect when configured with host & port

### ISSUE THREAD

**Issue by prwhitaker** [holland-backup/holland#210]
Holland 1.1.5 can't connect to MySQL when the backupset is configured to use a host & port, rather than connecting over a Unix socket

With the following in the backupset config:
```
[mysql:client]
user = mysql_backup
password = <password removed>
socket = /var/lib/mysql/mysql.sock
```
it works fine:
```
# holland bk --dry-run
Holland 1.1.5 started with pid 18018
--- Starting dry run ---
Creating backup path /var/spool/holland/default/20181003_174935
Estimating size of mysqldump backup
Estimated Backup Size: 2.46MB
Using Historic Space Estimate: Checking for information in /var/spool/holland/default/newest/backup.conf
Last backup used 342.60KB
The new backup estimate is at least 1.5 times the size of the currnet estimate: 2.46MB > (0.00B * 1.5). Default back to 'estimated-size-factor'
Adjusting estimated size by 0.50 to 1.23MB
Starting backup[default/20181003_174935] via plugin mysqldump
Running in dry-run mode.
Using mysqldump executable: /usr/bin/mysqldump
mysqldump version 5.7.22
Using mysqldump option --flush-privileges
Using mysqldump option --max-allowed-packet=128M
Using gzip compression level 1 with args 
Wrote backup manifest /var/spool/holland/default/20181003_174935/backup_data/MANIFEST.txt
Executing: /usr/bin/mysqldump --defaults-file=/var/spool/holland/default/20181003_174935/my.cnf --flush-privileges --max-allowed-packet=128M --lock-tables mysql
Executing: /usr/bin/mysqldump --defaults-file=/var/spool/holland/default/20181003_174935/my.cnf --flush-privileges --max-allowed-packet=128M --single-transaction sys
Backup completed in 0.06 seconds
--- Ending dry run ---
```

However, switching to a host/IP and port in the backupset config:
```
[mysql:client]
user = mysql_backup
password = <password removed>
host = 127.0.0.1
port = 3306
```
results in this error:
```
# holland bk --dry-run
Holland 1.1.5 started with pid 32173
--- Starting dry run ---
Creating backup path /var/spool/holland/default/20181003_175753
Estimating size of mysqldump backup
Failed to connect to database
an integer is required
MySQL Error an integer is required
Backup failed: MySQL Error an integer is required
--- Ending dry run ---
```

**Comment by soulen3** [comment]
Looks like this only effect holland when a port is defined in the config. I've got a fix for this in 1.1.7

**Comment by soulen3** [comment]
This was an issue with the port integer getting cased to a str in holland.lib.mysql. This was introduced in version 1.1.0 with adding python 3 compatibility. 

1.1.7 has been released.

**Comment by gsdevme** [comment]
```
holland --version
1.1.7
```

```
 yum list installed | grep holland
holland.noarch                    1.1.7-1.el6                      @epel
holland-common.noarch             1.1.7-1.el6                      @epel
holland-mysql.noarch              1.1.7-1.el6                      @epel
holland-mysqldump.noarch          1.1.7-1.el6                      @epel
```

@soulen3 what point release of 1.1.7, via epel el6 Im still seeing this. Did this get resolved for you @prwhitaker

**Comment by soulen3** [comment]
I thought this was fixed in 1.1.7, but maybe I missed something.  Can you please post you're error message here?

**Comment by gsdevme** [comment]
Hi, Im not seeing any error messages exactly aside from its not able to connect, we have no issues connecting via console on port 29300. Can holland outputs its connection information via some debug command?

This holland configuration was working on `1.0.14-3.el6` 

during some auto patching it was upgraded to 1.1.5 and then 1.1.7 but continues to be failing since the 3rd of October (the original update)

```
Oct 03 04:08:57 Updated: holland-1.1.5-1.el6.noarch
Oct 03 04:08:57 Updated: holland-common-1.1.5-1.el6.noarch
Oct 03 04:08:57 Updated: holland-mysql-1.1.5-1.el6.noarch
Oct 03 04:08:57 Updated: holland-mysqldump-1.1.5-1.el6.noarch
Oct 06 03:55:33 Updated: holland-1.1.7-1.el6.noarch
Oct 06 03:55:33 Updated: holland-common-1.1.7-1.el6.noarch
Oct 06 03:55:33 Updated: holland-mysql-1.1.7-1.el6.noarch
Oct 06 03:55:34 Updated: holland-mysqldump-1.1.7-1.el6.noarch
```

```
# define holland user information
user = xxxxxxx
password = xxxxxxx
host = 127.0.0.1
port = 29300
```

```
[root@dbvault backupsets]# mysql -h 127.0.0.1 -p -u xxxx -P 29300 xxxxx
Enter password:
Reading table information for completion of table and column names
You can turn off this feature to get a quicker startup with -A

Welcome to the MariaDB monitor.  Commands end with ; or \g.
Your MariaDB connection id is 8158064
Server version: 10.0.31-MariaDB-1~wheezy mariadb.org binary distribution

Copyright (c) 2000, 2018, Oracle, MariaDB Corporation Ab and others.

Type 'help;' or '\h' for help. Type '\c' to clear the current input statement.

MariaDB [xxxx]>
```

**Comment by soulen3** [comment]
It might not be showing the error message because of the logging level set in `holland.conf`.  Try adding `-l debug` to the holland command.

This could be related to https://github.com/holland-backup/holland/issues/213. What is `estimate-method` set to in you're backupset configuration?

**Comment by gsdevme** [comment]
I put holland in debug mode and re-ran however it doesn't really give me anymore output just mysql connection errors. 

`estimate-method` we don't have it defined so its whatever the default is

**Comment by gsdevme** [comment]
I've just downgraded to `1.0.10-1` for now and its restored holland to a working state.

**Comment by soulen3** [comment]
For reason that I haven't figured out yet, Holland is treating `localhost` and `127.0.0.1` differently.  Can you try it using `localhost` instead of the IP address?

```
# cat ~/.my.cnf
[client]
user=root
password=MyNewPass
host = 127.0.0.1
port = 29300
# holland bk
Unknown parameter 'format' in section 'logging'
Holland 1.0.14 started with pid 30050
--- Starting backup run ---
Creating backup path /var/spool/holland/default/20181010_212207
Estimating size of mysqldump backup
Estimated Backup Size: 641.84KB
Starting backup[default/20181010_212207] via plugin mysqldump
Using mysqldump executable: /usr/bin/mysqldump
mysqldump version 5.1.73
Using mysqldump option --flush-privileges
Using mysqldump option --routines
Using mysqldump option --events
Using mysqldump option --max-allowed-packet=128M
Using gzip compression level 1 with args
Wrote backup manifest /var/spool/holland/default/20181010_212207/backup_data/MANIFEST.txt
Executing: /usr/bin/mysqldump --defaults-file=/var/spool/holland/default/20181010_212207/my.cnf --flush-privileges --routines --events --max-allowed-packet=128M --lock-tables mysql
Executing: /usr/bin/mysqldump --defaults-file=/var/spool/holland/default/20181010_212207/my.cnf --flush-privileges --routines --events --max-allowed-packet=128M --single-transaction te
st
Final on-disk backup size 173.69KB
27.06% of estimated size 641.84KB
Backup completed in 0.13 seconds
Purged default/20181010_212010
1 backups purged
Released lock /etc/holland/backupsets/default.conf
--- Ending backup run ---
```

After updating Holland
```
# holland bk
Holland 1.1.7 started with pid 30243
--- Starting backup run ---
Creating backup path /var/spool/holland/default/20181010_212354
Estimating size of mysqldump backup
Failed to connect to database
(2003, "Can't connect to MySQL server on '127.0.0.1' (111)")
MySQL Error (2003, "Can't connect to MySQL server on '127.0.0.1' (111)")
Backup failed: MySQL Error (2003, "Can't connect to MySQL server on '127.0.0.1' (111)")
Released lock /etc/holland/backupsets/default.conf
--- Ending backup run ---
# sed -i 's/127.0.0.1/localhost/g' ~/.my.cnf
# holland bk
Holland 1.1.7 started with pid 30479
--- Starting backup run ---
Creating backup path /var/spool/holland/default/20181010_212542
Estimating size of mysqldump backup
Estimated Backup Size: 641.84KB
Using Historic Space Estimate: Checking for information in /var/spool/holland/default/newest/backup.conf
Last backup used 173.69KB
The new backup estimate is at least 1.5 times the size of the currnet estimate: 641.84KB > (0.00B * 1.5). Default back to 'estimated-size-factor'
Starting backup[default/20181010_212542] via plugin mysqldump
Using mysqldump executable: /usr/bin/mysqldump
mysqldump version 5.1.73
Using mysqldump option --flush-privileges
Using mysqldump option --routines
Using mysqldump option --events
Using mysqldump option --max-allowed-packet=128M
Using gzip compression level 1 with args
Wrote backup manifest /var/spool/holland/default/20181010_212542/backup_data/MANIFEST.txt
Executing: /usr/bin/mysqldump --defaults-file=/var/spool/holland/default/20181010_212542/my.cnf --flush-privileges --routines --events --max-allowed-packet=128M --lock-tables mysql
Executing: /usr/bin/mysqldump --defaults-file=/var/spool/holland/default/20181010_212542/my.cnf --flush-privileges --routines --events --max-allowed-packet=128M --single-transaction test
Final on-disk backup size 173.69KB
27.06% of estimated size 641.84KB
Backup completed in 0.11 seconds
Purged default/20181010_212354
Purged default/20181010_212207
2 backups purged
Released lock /etc/holland/backupsets/default.conf
--- Ending backup run ---
```

**Comment by prwhitaker** [comment]
@gsdevme the fix in 1.1.7 worked fine for me

@soulen3 the behavior you've just seen with localhost/127.0.0.1 is almost certainly due to the long-standing "feature" in mysql that `localhost` == connect via Unix socket, and ignore the TCP info entirely

Noted in the MySQL docs here: https://dev.mysql.com/doc/refman/8.0/en/connecting.html#option_general_protocol

Complained about at least 10 years ago:
https://bugs.mysql.com/bug.php?id=31577

**Comment by soulen3** [comment]
thanks @prwhitaker , that explains why switching to 'localhost' works. 

I'm still working on trying to figure out what changed caused this to break between 1.0.14 and 1.1.7
